### ISSUE TITLE
Talos grid search freezes after a few assays

### ISSUE THREAD

**Issue by TelmoAdao** [autonomio/talos#159]
Dear developers,

In advance, thank you very much for making available such useful tool. 

Although, I would like to report a strange behavior that I have been experiencing. When I try to use a grid to optimize SGD parameters (lr, decay and momentum - without grid downsample), Talos successfully starts to run several assays but after a few rounds, it just freezes indefinitely. I am using a workstation equipped with a dual Xeon Intel (54 logical cores, not quite sure the precise reference right now), 128 GB of RAM and two SLI paired Quadro M4000. Regarding the dataset, it is a short one, of around 300 pics divided by 6 classes, just for testing purposes. I coded according to your Notebook examples.

Would you please tell me what am I doing wrong? If you need more details, please, just let me know.

Best regards,
Telmo

**Comment by TelmoAdao** [comment]
Dear developers,

Best wishes for this new year!

Yes, i tried to run the examples with success. I was even able to use talos with a short grid of only two lines (fixed LR, fixed momentum and 2 decays for SGD fine-tunning) and everything went just fine. But when I try to expand the parameters increasing the grid to 27 combinations, talos freezes after the fourth or fifth combination (for testing purposes, I am using rounds of only two epochs). It's just weird, don't you think?

In advance, thank you for your support.

**Comment by mikkokotila** [comment]
Ok, then we start digging :) Can you share with a sample of your data (n=10 is enough), your input model, param grid, and the Scan() command. In case the data is sensitive ie. you can't share it, you could share dummy data with the same format where your model runs successfully without Talos.

**Comment by TelmoAdao** [comment]
@mikkokotila, here are the elements you asked for (please, find them attached)!

The main programming lines used to explore Talos features:
[Code.pdf](https://github.com/autonomio/talos/files/2732570/Code.pdf)
[Code.docx](https://github.com/autonomio/talos/files/2732571/Code.docx)

Eclipse console results (freezing at 25% after running a few rounds):
[Eclipse_Console_Result.txt](https://github.com/autonomio/talos/files/2732580/Eclipse_Console_Result.txt)

Used dataset sample (around 450 images for training + validation):
[Subset-Leaves-General.zip](https://github.com/autonomio/talos/files/2732581/Subset-Leaves-General.zip)

Thank you very much for your help!

**Comment by mikkokotila** [comment]
It seems that you are wrapping Talos experiment inside some kind of optimization function? Is that correct? What are you trying to achieve by doing this?

**Comment by TelmoAdao** [comment]
Yes, that's right. My aim is to obtain the best parameters among the ones shown in the code (LR, decay, momentum) to then return them as function result. I am aware the "return" instruction is missing; however, talos.Scan(...) frezees at 25% while running optimization over the dataset that I made available, thus preventing subsequent instruction of being executed.

**Comment by mikkokotila** [comment]
When you say "function" result, do you mean that you want to end up with a function that can do predictions with the best model? 

If that's the case, you can achieve the same (cross-validated "best" model and the resulting predictions) in two very simple ways with just two lines of code:

```
scan_object = ta.Scan(x=x_train , y=y_train, params=p, model=gen_model)
scan_object_with_eval = ta.Autom8(scan_object, x_val, y_val)
```
The above will entail completing an entire pipeline from the scan to evaluation and predictions. The prediction and evaluation results will be added to the scan_object. 

For the most part, you might just want to be able to evaluate the models right there and then and not think about predictions yet, and that's when you can just simply:

```
scan_object = ta.Scan(x=x_train , y=y_train, params=p, model=gen_model)
scan_object.evaluate_models(x_val, y_val)
```
Would that do the job for you?

**Comment by TelmoAdao** [comment]
@mikkokotila, i did not have the opportunity to test it yet, but looks good. I will try your suggestion as soon as possible and I'll let you now the results. Thank you ;)

**Comment by TelmoAdao** [comment]
I tried your suggestion but the outcome was the same... Talos scan stucks in the 25% of progress :S Should I test something else?

**Comment by mikkokotila** [comment]
It depends on the OS you are on, but you can try: 

`ta.Scan(...  clear_tf_session=False  ...)`

Generally speaking, also looking at Keras issues where kernels silently die, the issue is caused by resource management (like running out of memory etc). Also some users report that kernel dying is fixed by updating conda (are you on conda?) and/or updating TensorFlow. Also updating cudNN have been reported to help. Can you try those and see what happens.

**Comment by TelmoAdao** [comment]
Wow! It worked! Thanks a lot!

Although, the "evaluate_models" method mencioned earlier is not working. Here is the result: 
  File "...\ML\MLTools.py", line 198, in gridSearch_Talos_SGD_TimeBasedShedule
    all_done = talos.Scan(x=X , y=Y, params=p, model=create_model, clear_tf_session=False).evaluate_models(Xval, Yval)
TypeError: func_evaluate() missing 1 required positional argument: 'y_val'

It looks like Scan method requires a 'y_val' parameter. I've also noticed that "evaluate_models" invocation seems to not be supported. Maybe, the object returned by talos.Scan does not have that definition...

What do you think?

Thank you very much for your help!
